### PR TITLE
API: Group CRUD

### PR DESCRIPTION
**Pull request by HugoGresse**
- groups add/edit and members management
- new collection,s within than splitted members with UID and membersDetails as `id: {id, name`. This improve where operation to use arrayContains diretly within the search operand. 
- some http code refacto (to be used later on other pr)

missing piece to be added later on: 
- subscribeToGroup with registration token
- invites via push notif

### CODE REVIEW

File: functions/src/api/v1/group/createGroup.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    const name = req.body.name
+    const members = JSONParse(req.body.members || '[]')
+
+    const groupId = await GroupDao.add(userId, name, members)
+    await GroupDao.update({
+        id: groupId,
+    })
+
+    res.send({
```
Comment (ntchou) [inline functions/src/api/v1/group/createGroup.ts:57]: missing 201 http status code

File: functions/src/api/v1/group/editGroup.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ *     produces:
+ *     - application/json
+ *     parameters:
+ *       - name: name
+ *         description: The group name
+ *         in: x-www-form-urlencoded
+ *         required: true
+ *         type: string
+ *     responses:
+ *       200:
```
Comment (ntchou) [inline functions/src/api/v1/group/editGroup.ts:26]: perhaps`204` instead `200`

File: functions/src/api/v1/group/editGroup.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    await assertGroupEditAllowed(userId, groupId)
+
+    const name = req.body.name
+
+    await GroupDao.update({
+        id: groupId,
+        name,
+    })
+
+    res.status(200).send()
```
Comment (ntchou) [inline functions/src/api/v1/group/editGroup.ts:49]: see my comment above

File: functions/src/api/v1/group/groupRights.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import { GroupDao } from '../../../db/GroupDao'
+
+export const assertGroupEditAllowed = async (userId: UID, groupId: GroupId) => {
+    const group = await GroupDao.get(groupId)
+    if (group.ownerUserId !== userId) {
+        throw new GroupEditForbiddenError()
+    }
+}
+
+export const assertGroupReadAllowed = async (userId: UID, groupId: GroupId) => {
```
Comment (ntchou) [inline functions/src/api/v1/group/groupRights.ts:16]: missing return type

File: functions/src/db/GroupDao.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    public static async add(
+        ownerUserId: UID,
+        name: string,
+        members?: Member[]
+    ): Promise<GroupId> {
+        const documentReference = await db.collection(COLLECTIONS.groups).add({
+            name,
+            ownerUserId,
+            members: members ? members.map((member) => member.id) : [],
+            membersDetails: members ? members : {},
```
Comment (ntchou) [inline functions/src/db/GroupDao.ts:73]:
```suggestion
            membersDetails: members ||  {},
```

File: functions/src/types/Group.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+export interface Group {
+    id: GroupId
+    name: string
+    ownerUserId: UID
+    members: UID[]
+    createdAt: Timestamp
+    updatedAt: Timestamp
+}
+
+export interface PubicGroup {
```
Comment (ntchou) [inline functions/src/types/Group.ts:15]:
```suggestion
export interface PublicGroup {
```

**Review by ntchou — COMMENTED** [review]
Good job

File: functions/src/api/v1/group/addMembersToGroup.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    const userId: UID = res.locals.uid
+    const groupId: GroupId = req.params.groupId
+    await assertNewResourceCreationGranted(userId)
+    await assertGroupEditAllowed(userId, groupId)
+
+    const members = JSONParse(req.body.members || '[]')
+
+    await GroupDao.addMembers(groupId, members)
+
+    res.status(200).send()
```
Comment (MattiaPrimavera) [inline functions/src/api/v1/group/addMembersToGroup.ts:47]: `204` for `no-content` response ?

File: functions/src/api/v1/group/getGroup.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import { UID } from '../../../types/uid'
+import { GroupId } from '../../../types/Group'
+import { assertGroupReadAllowed } from './groupRights'
+import { GroupDao } from '../../../db/GroupDao'
+
+/**
+ * @swagger
+ * /v1/groups/{groupId}/:
+ *   get:
+ *     description: Get the given group. Members user ids will be returned for regular members and all members details will be returned for the owner.
```
Comment (MattiaPrimavera) [inline functions/src/api/v1/group/getGroup.ts:12]: shouldn't only the `owner` be able to see the members list ?

File: functions/src/api/v1/group/removeMembersFromGroup.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ *       403:
+ *         $ref: '#/components/responses/403'
+ *       412:
+ *         $ref: '#/components/responses/412'
+ */
+export const removeMembersFromGroup = wrap(
+    async (req: Request, res: Response) => {
+        const userId: UID = res.locals.uid
+        const groupId: GroupId = req.params.groupId
+        await assertNewResourceCreationGranted(userId)
```
Comment (MattiaPrimavera) [inline functions/src/api/v1/group/removeMembersFromGroup.ts:41]: why this one here ?

File: functions/src/db/GroupDao.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+export interface GroupEditData {
+    id: GroupId
+    name?: string
+}
+
+export interface DatabaseGroup extends Group {
+    membersDetails: Member[]
+}
+
+export class GroupDao {
```
Comment (MattiaPrimavera) [inline functions/src/db/GroupDao.ts:23]: pls consider creating a generic `Dao` in order to factor basic CRUD operations in a single spot

**Review by MattiaPrimavera — CHANGES_REQUESTED** [review]

File: functions/src/api/v1/group/getGroup.ts (hunk 8 above)
Comment (HugoGresse) [reply functions/src/api/v1/group/getGroup.ts:12]: it is, wrong desc ty

File: functions/src/db/GroupDao.ts (hunk 10 above)
Comment (HugoGresse) [reply functions/src/db/GroupDao.ts:23]: I will consider it :D

**Pull request merged**